### ISSUE TITLE
Persistable semanticdb search index

### ISSUE THREAD

**Issue by olafurpg** [scalameta/language-server#35]
The server currently re-builds a semanticdb index on every startup. This is fine for small projects, but is not scalable for larger projects. Also with #23, indexing will always take at least ~10-15s since we now index the sources of your full classpath, including the JDK (2.5M loc). If we persist the semanticdb index, the language server only needs to run this expensive computation once and startup will be instant from there on.

There are many alternatives to store these indices

- protobuf, we already store semanticdbs as protobuf messages. We could add more messages to the schema to represent pre-computed indices like "references to symbol". In Metadoc, we use this technique to generate a static site with IDE-like features http://scalameta.org/metadoc/ See the schema https://github.com/scalameta/metadoc/blob/6368d632cb717883d63c56b341775e5f51c1edbd/metadoc-core/src/main/protobuf/metadoc.proto This approach is perfect as long as our schema stayes simple, we can fit the entire index into memory, and out queries remain simple.
- sqlite, we can store the indices in a sqlite database, one db for each workspace. The downside of this approach is that we bring in a piece of technology into our stack. This approach will however work great if our schema becomes more complex (example, support multiple snapshots of the same file from different git checkouts) and the index becomes too big to fit into memory.

For now, I think we should go with the protobuf approach since I believe that our schema can and should remain dead simple and the size of the indices stay small enough to fit in memory. On server startup, we load a `.metaserver/index.zip` file (filename up for debate). On server shutdown, we optionally dump the index to disk for use next time we open the editor.

I propose we make minor adaptations to the metadoc schema
```proto
syntax = "proto3";

package scala.meta.languageserver.index;

// All of the metadata associated with a single symbol definition
message SymbolIndex {
    string symbol = 1;
    // The Position where this symbol is defined.
    Position definition = 2;
    map<string, Ranges> references = 3;
    int64 flags = 4;
    string name = 6;
    string signature = 7;
    // Planned for Scalameta v2.2, see https://github.com/scalameta/scalameta/milestone/9
    // string docstring = 6; // javadoc/scaladoc
    // string overrides = 7; // parent symbol that this symbol overrides
}

// A standalone position used by "Go to definition" when jumping to a symbol
// definition.
message Position {
  // can be a local file, entry inside local zip file, http url, or anything
  // else.
  string uri = 1;
  int32 start = 2;
  int32 end = 3;
}

// A hack to workaround the fact that protobuf can't 
// encode `map<string, repeated string>`
message Ranges {
  repeated Range ranges = 1;
}

// A slim range position, the filename is embedded in the
// key of `references: map<string, ranges>`
message Range {
  int32 start = 2;
  int32 end = 3;
}
```
Observe that the index does not include the uri file contents or the original semanticdb `Document`, since that would be prohibitive to store in memory.
The language server must be able to retrieve the original semanticdb `Document` from a URI, which includes the full file contents and all of the resolved names, symbols and synthetics.

**Comment by olafurpg** [comment]
BTW, we can probably reuse the existing indexer for this schema from Metadoc https://github.com/scalameta/metadoc/blob/6368d632cb717883d63c56b341775e5f51c1edbd/metadoc-cli/src/main/scala/metadoc/cli/MetadocCli.scala#L164-L200 It's fairly well optimized, my local experiments show that it can index a corpus semanticdbs from ~3m LOC in ~5-10s on my laptop.

**Comment by ShaneDelmore** [comment]
I like this idea in general, but so far have found when working with these case classes that they are missing logic that makes them convenient to work with.  A simple example is the hasFlags function, we end up converting the case class to a different case class to add functionality to it. Given this is a case where we really do care about performance I would suggest we move the functionality that is currently on dedicated case classes into standalone functions that operate on the case classes, or traits that are mixed into the generated classes.  This would prevent us from wanting to convert protobuf.Denotation into semanticdb.Denotation to figure out if it is a primary constructor, for instance.

**Comment by olafurpg** [comment]
For operations on the generated classes we can add extension methods or use static methods where we pass in the generated case classes instances. `HasFlags` can be mixed into the extension method, if we make the `trait HasFlags extends Any` then we can make the extension methods AnyVals.

There are not many utility methods we need to build the index, at least I didn't find need for them in the metadoc indexer.

**Comment by olafurpg** [comment]
Note, I just updated the proposed schema to remove `Denotation` by inlining it into SymbolIndex.

**Comment by olafurpg** [comment]
Should we maybe store line number + column instead of offsets in this index? LSP only operates with line+column, so I think it makes sense to delegate that we do the same. The indexer can do the work of converting offsets to line+column.

**Comment by ShaneDelmore** [comment]
I’m unsure about this.  I actually wish lsp took absolute offsets and translated them client side.  Editors have all files loaded so line/column seems a decent way to go but for other uses it seems quite a bit less efficient.  For example, if we have the ability to “show definition” of a method, and the contents of the file, with absolute position we could index directly into the correct range, with line and column we have to read every byte looking for new lines.

**Comment by ShaneDelmore** [comment]
Do you plan to store one giant index file, or multiple files?  I was thinking of possibly storing a file per lib/version which would make it easy to reuse for example jdk1.8.blah for all projects on a developers machine if they all use the same jdk and prevent needing to rebuild it for each one.

**Comment by olafurpg** [comment]
>  I actually wish lsp took absolute offsets and translated them client side. 

I agree, but I don't think there is much we can do about this https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#text-documents
By storing line+column in our index, we avoid having to fetch the source file contents of all files that reference a particular symbol, which is a win.

> Do you plan to store one giant index file, or multiple files?

I was thinking about how to handle this. I was thinking of storing it as one large compressed file, something like `.metaserver/symbols.zip` which is a compressed `SymbolCorpora`

```proto
message SymbolCorpora {
  repeated SymbolCorpus corpora = 1;
}

message SymbolCorpus {
  string uri = 1; // uniquely identifies a corpus, example "file://jdk.jar"
  repeated SymbolIndex symbols = 2;
}
```

Names are up for debate. However, I'm not so sure how much caching by jar makes sense, since the contents of `references` is dependent on your project sources.

**Comment by ShaneDelmore** [comment]
> Names are up for debate. However, I'm not so sure how much caching by jar makes sense, since the contents of references is dependent on your project sources.

I don't understand the relevance.  Are you saying the cache would be kept in references?

**Comment by olafurpg** [comment]
Fixed in #45
